modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -167,31 +203,57 @@ void run(ClusterState state) {
+             * exclude it from force merging later in this DLM run.
+             */
+            Index originalWriteIndex = dataStream.getWriteIndex();
             try {
                 maybeExecuteRollover(state, dataStream);
             } catch (Exception e) {
                 logger.error(() -> String.format(Locale.ROOT, "DLM failed to rollver data stream [%s]", dataStream.getName()), e);
                 DataStream latestDataStream = clusterService.state().metadata().dataStreams().get(dataStream.getName());
                 if (latestDataStream != null) {
-                    if (latestDataStream.getWriteIndex().getName().equals(writeIndex)) {
+                    if (latestDataStream.getWriteIndex().getName().equals(originalWriteIndex.getName())) {
                         // data stream has not been rolled over in the meantime so record the error against the write index we
                         // attempted the rollover
-                        errorStore.recordError(writeIndex, e);
+                        errorStore.recordError(originalWriteIndex.getName(), e);
                     }
                 }
             }
-
+            Set<Index> indicesBeingRemoved;
             try {
-                maybeExecuteRetention(state, dataStream);
+                indicesBeingRemoved = maybeExecuteRetention(state, dataStream);
             } catch (Exception e) {
+                indicesBeingRemoved = Set.of();
                 // individual index errors would be reported via the API action listener for every delete call
                 // we could potentially record errors at a data stream level and expose it via the _data_stream API?
                 logger.error(
                     () -> String.format(Locale.ROOT, "DLM failed to execute retention for data stream [%s]", dataStream.getName()),
                     e
                 );
             }
+
+            try {
+                /*
+                 * When considering indices for force merge, we want to exclude several indices: (1) We exclude the current write index
+                 * because obviously it is still likely to get writes, (2) we exclude the most recent previous write index because since
+                 * we just switched over it might still be getting some writes, and (3) we exclude any indices that we're in the process
+                 * of deleting because they'll be gone soon anyway.
+                 */
+                Set<Index> indicesToExclude = new HashSet<>();
+                Index currentWriteIndex = dataStream.getWriteIndex();
+                indicesToExclude.add(currentWriteIndex);
+                indicesToExclude.add(originalWriteIndex); // Could be the same as currentWriteIndex, but that's fine
+                indicesToExclude.addAll(indicesBeingRemoved);
+                maybeExecuteForceMerge(state, dataStream, indicesToExclude);
```

Comment:
This is a super neat trick (running it after retention essentially spares us the need for an "end of operation" marker - at least until we have operations following forcemerge). Very nice 🚀  ❤️ !